### DISCUSSION TITLE
GIT_SSH_COMMAND + Deploy Key = Permission Denied?

### DISCUSSION THREAD

**Discussion by andreas-github** [community/community#26692] [Actions]
<p>I am experimenting a little to familiarise myself with Deploy Keys and am not able to figure out why the following is not working:</p>
<pre><code class="lang-auto">root@host:~# GIT_SSH_COMMAND='ssh -i key -o IdentitiesOnly=yes -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no'
root@host:~# git clone git+ssh@github.com:*github-user*/*github-repo*.git
Cloning into '*github-repo*'...
git+ssh@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
</code></pre>
<ul>
<li>I have added the public key as a Deploy Key read-only) to the <em>github-repo</em>.</li>
<li>The <em>key</em> file contains the matching private key.</li>
<li>
<em>github-user</em> is the maintainer of <em>github-repo</em> and has ssh keys as well.</li>
<li>the host is a completely freshly installed Ubuntu 20.04. with <strong>no</strong> git or ssh operations done before this example.</li>
<li><code>git version 2.25.1</code></li>
</ul>
<p>Why does this not work? Why the permission error?</p>
<p>Grateful for any thoughts and ideas what the issue is here. Thanks.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="andreas-github" data-post="1" data-topic="248353">
<div class="title">
<div class="quote-controls"></div>
 andreas-github:</div>
<blockquote>
<p><code>git+ssh@github.com</code></p>
</blockquote>
</aside>
<p><code>git+ssh</code> is not a valid SSH user. To access GitHub over SSH you have to use the <code>git</code> user.</p>

**Comment by andreas-github** [comment]
<aside class="quote group-Advisory-Council" data-username="airtower-luna" data-post="2" data-topic="248353">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p><code>git+ssh</code> is not a valid SSH user. To access GitHub over SSH you have to use the <code>git</code> user.</p>
</blockquote>
</aside>
<p>Ok, changed it to <code>git clone git@github.com:*github-user*/*github-repo*.git .</code> but the error is still exactly the same :slightly_frowning_face:</p>
<p>Additional detail about this test:</p>
<ul>
<li>the repo is a private one</li>
<li>the GitHub account used is on a FREE plan</li>
</ul>

**Comment by airtower-luna** [comment, marked as answer]
<p>I notice you don’t <code>export</code> the <code>GIT_SSH_COMMAND</code> variable, so it’s not actually available to the subsequent <code>git</code> command. That’s probably the issue.</p>
<p>However, before you continue, I see a number of things there that practically yell <em>really bad idea</em>. :warning:</p>
<ul>
<li>You’re running Git as root. Git doesn’t do anything that needs root access, and running anything as root unnecessarily is an unnecessary security risk.</li>
<li>You disable the SSH known hosts file and host key checking. This means that anyone who can mess with your network traffic can make you connect to anywhere they want, and have your Git command download whatever they want.</li>
<li>I assume that if you clone in root, you intend to use the repository as root. In combination with not checking the identity of the remote host that makes it really easy for an attacker to get remote code execution.</li>
</ul>
<p>Get <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints" class="inline-onebox">GitHub's SSH key fingerprints - GitHub Docs</a>, retrieve the keys, verify, and add them to the right user’s known hosts file, and use that file when connecting.</p>
<p>Don’t clone as root, clone as a regular user. If your tool really needs to run (or install) as root you can still do that after. Good luck! :slightly_smiling_face:</p>

**Comment by andreas-github** [comment]
<p>Thanks for your help <a class="mention" href="/airtower-luna">@airtower-luna</a></p>
<p>It was indeed the missing <code>export</code> for <code>GIT_SSH_COMMAND</code> :persevere:</p>
<p>Regarding <code>root</code>, that is just because this was originally a quick test in VPS instance using the default wb console login (quick &amp; dirty)…I 100% subscribe to your advice.</p>
<p>For what it’s worth, here is the “working” shell snippet (w/o the <em>bad</em> parts :slight_smile:</p>
<pre><code class="lang-auto">user@host:~# export GIT_SSH_COMMAND='ssh -i key -o IdentitiesOnly=yes'
user@host:~# git clone git@github.com:*github-user*/*github-repo*.git .
Cloning into '.'....

</code></pre>
